compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -675,8 +675,22 @@ fn construct_error(tcx: TyCtxt<'_>, def_id: LocalDefId, guar: ErrorGuaranteed) -
                         ))),
                     )
                 }
-                _ => {
-                    span_bug!(span, "expected type of closure body to be a closure or coroutine");
+                ty::CoroutineClosure(did, _args) => {
+                    let inputs = tcx
+                        .closure_user_provided_sig(did.expect_local())
+                        .value
+                        .skip_binder()
+                        .inputs();
+
+                    let err = Ty::new_error(tcx, guar);
+                    (inputs.iter().map(|_| err).collect(), err, None)
+                }
```

Comment:
@compiler-errors not pretty, but it avoids any issues downstream.

Reply:
:ferrisclueless: can u add a `FIXME(async_closures)` here? I will elaborate it with the proper generated signature.

Reply:
if you want, but there's not supposed to be anything useful coming out of this mir body anyway, so not sure it would ever effect anything, while being more complex

Reply:
If it really doesn't matter, then perhaps the changes in #117418 should be scrapped and this whole body should be simplified back to what it used to do, which is just emit `ty::Error` for all inputs/outputs.

Reply:
It doesn't matter right now. I guess it may matter in the future. We can always scrap it later if it turns out to never improve diagnostics or ICE resistance